### DISCUSSION TITLE
How to hide query params while using dynamic routes?

### DISCUSSION THREAD

**Discussion by kouliavtsev** [vercel/next.js#18732] [Help] [answered]
Hi,

On the homepage, there is a list component that has a list item that looks like this.

```
import Link from 'next/link';

const Card = ({ id, moodType }) => {
  return (
    <Link as={`/moods/${moodType}?id=${id}`} href={`/moods/[slug]`}>
      <a>{mood}</a>
    </Link>
  );
};

export default Card;
```
To keep it simple there are two moodTypes (either, sad or happy).

The folder structure looks like this:
```
cat
  moods
    [slug].js
```

When the user clicks on a link he gets a link looking like this:
`http://localhost:3000/moods/happy?id=SkfA2GcedP`

The id is passed to fetch all the data that is related to the moodType.

I am new to nextjs and I wonder how can I make the URL look cleaner, looking like this `http://localhost:3000/moods/happy`, while still passing the id to the next component.

Thanks a bunch!
Anton

**Comment by jamesmosier** [comment, marked as answer]
Hi @kouliavtsev. There is no way to hide any part of the url with Next unfortunately or to pass data between pages. Most routes I see with subcategories are formatted like `pages/moods/[mood]/[id]`.

What's the difference between `mood` and `id` in your use case?

**Reply by kouliavtsev** [reply]
Hey @jamesmosier, thanks for clearing that up.  :)

I have solved the issue by creating an additional field in the api `slug` (uid). Which made it possible to have pretty urls on the next page and also made it possible to fetch data from api.

**Reply by fkoner** [reply]
@kouliavtsev How did you solve this? Do you have an example?

**Reply by hjvasoya009** [reply]
@kouliavtsev 
I'm facing the same issue. Can you share your answer in detail or some example of how you solved it please?

**Reply by alejo4373** [reply]
This possible with `as` prop for `<Link>` and the `as` param for `router.push()` https://www.reddit.com/r/nextjs/comments/j5dkzg/how_to_visit_url_and_hide_its_params_in_browser/

**Reply by Vishakh-vcc** [reply]
@kouliavtsev @hjvasoya009 @fkoner @pascuflow 
can anyone plz suggest me how did you all solve this issue...I am facing the same problem now of removing query params while dynamic routing using Router.push()

**Comment by andres-dejesus** [comment]
Is there still no way to do this? I need to auth certain pages in front end navigation that use Link and don't want users to see the token as a param, I have _middleware.ts, tried everything under the sun:

`  if (req.nextUrl.searchParams.get('token')) {
      return NextResponse.redirect('req.searchParams.delete('token').cookie('token', '') //Tried rewrite too
}`
Front end Link looks like this:    
`<Link href={{pathname:"/user/edit-profile", query:{token:token}}} activeClassName="active">`
This does not work either:
`<Link href={`/user/edit-profile?token=${token}`} as={'/user/edit-profile'} replace={true} activeClassName="active">`

**Reply by andres-dejesus** [reply]
Only workaround I found was to put the token on the cookie and access it on the _middleware.ts as req.cookies.token instead of a query param, I guess the direction moving forward for those ugly URL's is to have a fat cookie.
